### ISSUE TITLE
Include route's source IP in types.Routes returned from types.Result

### ISSUE THREAD

**Issue by sboeuf** [containernetworking/cni#496]
Right now, the structure Route https://github.com/containernetworking/cni/blob/v0.6.0/pkg/types/types.go#L115-L118 only includes the destination `Dst` and the gateway `GW`. It would be good to have also `Src` for more details about the route.
Basically changing this:
```go
type Route struct {
        Dst net.IPNet
        GW  net.IP
}
```
to this:
```go
type Route struct {
        Src net.IP
        Dst net.IPNet
        GW  net.IP
}
```

**Comment by sboeuf** [comment]
Is there anyone looking at issues on this repo ? I'd like to have a feedback about this one.

**Comment by rosenhouse** [comment]
@sboeuf Could you share a bit more about your use case?  Are you asking for the route structure in the [specification](https://github.com/containernetworking/cni/blob/master/SPEC.md#result) to change as well, or just the Go library?

**Comment by sboeuf** [comment]
We are relying on the Route structure here: https://github.com/containers/virtcontainers/blob/master/network.go#L67. We thought it was a good idea to rely on your generic structures to create our network results when scanning the underlying network. We use that both for our CNI and CNM implementation, regardless of the CNI relation.
To answer your question about what I need, basically the answer is only the Route structure to be modified to add this new field. But I also think that if we do that, any consumer of the library will expect the IPAM plugin to return results including `Src`. Unless that's okay to return `Src` being `nil`.

**Comment by rosenhouse** [comment]
@sboeuf I don't think we're interested in adding fields to the CNI Go library that aren't part of the CNI spec. 

Perhaps you can use a [structure embedding](https://golang.org/doc/effective_go.html#embedding) to address your need?

**Comment by sboeuf** [comment]
@rosenhouse Yes I could do that, but I wanted to check with you first. I don't understand why it is not part of the CNI spec, I mean when you describe a route, it can have a gateway, a destination, a source, a device, and some flags. Why are you not interested in adding those info to your Route structure ?

**Comment by rosenhouse** [comment]
@containernetworking/cni-maintainers I recall there being some prior discussion of adding a source field in the CNI spec, but couldn't find it....

**Comment by bboreham** [comment]
My understanding is the design of CNI IPAM mirrors that of DHCP. DHCP has no source field, therefore neither does CNI. 

I'd need to hear a use-case where specifying the source makes sense in a CNI context before voting to add it. What would the source be, apart from the address you just allocated?

**Comment by sboeuf** [comment]
It is not specific to the IPAM plugin, it's about the results that are returned after you've added a new network. My understanding is that after you call into `AddNetwork()` you can expect to get a full and detailed result describing your network. For that reason, I thought that adding `Src` or other fields could have been a good idea to describe better the underlying network.
